### ISSUE TITLE
Make it possible to run without using window.eval()

### ISSUE THREAD

**Issue by Jermolene** [Jermolene/TiddlyWiki5#6839]
[Content security policies](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) are a relatively new way to increase the security of web applications by having the server request additional security restrictions to be placed on a site.

In particular, it is now generally best practice to run web applications with the "unsafe-eval" restriction, which prevents the use of [window.eval()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval). That's a problem for TW5 because our built-in module system uses eval to execute modules.

An approach that should work in principle is to optionally:

* Preload all the required modules using `$tw.modules.define(moduleName,moduleType,exports)`, where "exports" is the object exported by the module
* Disable loading of modules from tiddlers
* We'd also need a slightly different build process to output the required modules in the required format

**Comment by pmario** [comment]
Isn't this a problem for every library that uses a "require.js" like module system?

**Comment by Jermolene** [comment]
> Isn't this a problem for every library that uses a "require.js" like module system?

I don't think so: modern tools build everything into one blob of JS as a build step on the server.

**Comment by buggyj** [comment]
Does this mean putting all js tiddlers inside a js script tag instead of in the json script tag?

**Comment by Jermolene** [comment]
> Does this mean putting all js tiddlers inside a js script tag instead of in the json script tag?

Yes just that. I'll edit the OP to clarify

**Comment by Leilei332** [comment]
As suggested on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_direct_eval!), it's better to use indirect eval or [Function constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function) (though still restricted by CSP, they are safer and faster).

**Comment by Jermolene** [comment]
Hi @Leilei332 I'd welcome further work on this. Do you think you could look at it in the v5.4.0 timescale?
